flow/aggregates.py

```diff
@@ -0,0 +1,463 @@
+            return True
+
+        return False
+
+    def __hash__(self):
+        blob_l = list(map(hash, [self._sort_by, self._reverse_order]))
+        # Get unique id for _aggregator and _select attributes
+        blob_l += list(map(self._get_unique_function_id, [self._aggregator, self._select]))
+        blob = ','.join(str(attr) for attr in blob_l)
+
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+    def _get_unique_function_id(self, func):
+        try:
+            return func.__code__.co_code
+        except Exception:  # Got partial function
+            return str(hash(func))
+
+    def _create_AggregatesStore(self, jobs):
+        """Create the instance of _AggregatesStore or _DefaultAggregateStore classes
+        to make sure that we don't explicitly store the aggregates in the FlowProject.
+
+        When iterated over these classes, an aggregate is yielded.
+        """
+        if not self._is_aggregate:
+            return _DefaultAggregateStore(jobs)
+        else:
+            return _AggregatesStore(self, jobs)
+
+    def __call__(self, func=None):
+        if callable(func):
+            setattr(func, '_flow_aggregate', self)
+            return func
+        else:
+            raise TypeError('Invalid argument passed while calling '
+                            'the aggregate instance. Expected a callable, '
+                            f'got {type(func)}.')
+
+
+class _AggregatesStore:
+    """This class holds the information of all the aggregates associated with
+    a :py:class:`Aggregate`.
+
+    This is a callable class which, when called, generates all the aggregates.
+    When iterated through it's instance, all the aggregates are yielded.
+
+    :param aggregate:
+        Aggregate object associated with this class.
+    :type aggregate:
+        :py:class:`Aggregate`
+    :param jobs:
+        The signac job handles
+    :type jobs:
+        Iterator
+    """
+    def __init__(self, aggregate, jobs):
+        self._aggregate = aggregate
+
+        # We need to register the aggregates for this instance using the
+        # jobs provided.
+        self._aggregates = list()
+        self._aggregate_ids = dict()
+        self._register_aggregates(jobs)
+
+    def __iter__(self):
+        yield from self._aggregates
+
+    def __getitem__(self, id):
+        "Return an aggregate "
+        try:
+            return self._aggregate_ids[id]
+        except KeyError:
+            raise LookupError(f'Unable to find the aggregate having id {id} in '
+                              'the FlowProject')
+
+    def __contains__(self, aggregate):
+        """Return whether an aggregate is stored in the this
+        instance of :py:class:`_AggregateStore`
+
+        :param aggregate:
+            Aggregate of jobs
+        :type jobs:
+            tuple of :py:class:`signac.contrib.job.Job`
+        """
+        if aggregate in self._aggregates:
+            return True
+        else:
+            return False
```

Comment:
I would prefer if we used the id to check for inclusion. It would be constant time rather than linear then.